hail/python/hailtop/aiogoogle/client/storage_client.py

```diff
@@ -0,0 +1,204 @@
+from typing import Tuple, Any, Set, Optional
+import asyncio
+import urllib.parse
+import aiohttp
+from hailtop.aiotools import AsyncStream, AsyncFS, FeedableAsyncIterable
+from multidict import CIMultiDictProxy
+from .base_client import BaseClient
+
+
+class InsertObjectStream(AsyncStream):
+    def __init__(self, it, request_task):
+        super().__init__()
+        self._it = it
+        self._request_task = request_task
+        self._value = None
+
+    def writable(self):
+        return True
+
+    async def write(self, b):
+        assert not self.closed
+        await self._it.feed(b)
+        return len(b)
+
+    async def _wait_closed(self):
+        await self._it.stop()
+        async with await self._request_task as resp:
+            self._value = await resp.json()
+            print(self._value)
+
+
+class GetObjectStream(AsyncStream):
+    def __init__(self, resp):
+        super().__init__()
+        self._resp = resp
+        self._content = resp.content
+
+    def readable(self) -> bool:
+        return True
+
+    async def read(self, n: int = -1) -> bytes:
+        assert not self._closed
+        return await self._content.read(n)
```

Comment:
Is the response keeping track of where it is in the stream and you don't manually have to do anything? If so, that's nice.

Reply:
Yep.  So `content` is a AsyncStream-like interface that's part of aiohttp, StreamReader:

https://docs.aiohttp.org/en/stable/streams.html#aiohttp.StreamReader

The Python standard library has something similar which is returned when creating network connections, StreamReader:

https://docs.python.org/3/library/asyncio-stream.html#asyncio.StreamReader

Annoyingly, neither of those interfaces can be implemented by external code, hence I created AsyncStream.   GetObjectStream is just a simple AsyncStream wrapper around the aiohttp stream.